### DISCUSSION TITLE
Compiling ort on macos M1 by linking library with system strategy / dynamic_load fails

### DISCUSSION THREAD

**Discussion by leichak** [pykeio/ort#29] [General]
Hey, i've been trying to build ort with pre-built by me onnxruntime 1.4.1 libraries and do it with mac m1. 
Although in both ways ort finds .dylib file, it prompts error because it says it wants x86_64. 
Is there a way to force it to want macos arm_64 instead? 

Daniel

**Comment by decahedron1** [comment]
What error is it giving?

**Reply by leichak** [reply]
For dynamic_load it prompts:

`thread 'tests::use_ort' panicked at 'could not load the library at `../../libs/libonnxruntime.dylib`: DlOpen { desc: "dlopen(../../libs/libonnxruntime.dylib, 0x0005): tried: \'../../libs/libonnxruntime.dylib\' (mach-o file, but is an incompatible architecture (have \'arm64\', need \'x86_64\')), ...`

**Reply by decahedron1** [reply]
Hmm, maybe the Rust code was built for `x86_64-apple-darwin`? Could you share the output of `rustup target list --installed`?

**Reply by leichak** [reply]
Yeah, seems like.

`x86_64-apple-darwin`

**Reply by leichak** [reply]
So I need to reinstall to arm64 as I understand?, or somehow force it to cross-compile

**Reply by decahedron1** [reply]
Yes, you can either install an arm64 version of rustup to use `aarch64-apple-darwin` as the default target, or add the targetwith `rustup target add aarch64-apple-darwin`, though that would mean you'd have to compile every Rust project with `--target=aarch64-apple-darwin`

**Reply by leichak** [reply]
Thank you it builds but then triplet is not supported in build.rs

`--- stderr
  thread 'main' panicked at 'Unsupported prebuilt triplet: MacOs, Arm64, None. Please use ORT_STRATEGY=system and ORT_LIB_LOCATION=/path/to/onnxruntime', /Users/.../build.rs:346:17`
  
  And indeed it is not.

**Reply by decahedron1** [reply]
> i've been trying to build ort with pre-built by me onnxruntime 1.4.1 libraries

If you want to link ort to your custom binaries you need to use either the `system` strategy or the `load-dynamic` feature.

Otherwise, if you want to use Microsoft binaries, Microsoft does actually provide prebuilt binaries for macOS arm64 and that error is a bug which I'll fix shortly

**Reply by decahedron1** [reply]
@leichak which crate are you using? `ort` supports macOS aarch64 since 1.13.0 (the first release), and `ort`'s error message is `Microsoft does not provide ONNX Runtime downloads for triplet`, not `Unsupported prebuilt triplet`...

**Reply by leichak** [reply]
Fair point, Thanks! I had onnxruntime-rs in deps as well and it was messing up stuff, without it with it worked.

`--target aarch64-apple-darwin` + prebuilt binary for arm64 + dynamic_load worked works.

In the future I will build some universal script building and loading depending on architecture. I will post it here for ppl.
